File: pronacej-backend-seguridad/src/main/java/net/latinus/sistema/integral/gestion/seguridad/service/IA/ActaExternamientoService.java
package net.latinus.sistema.integral.gestion.seguridad.service.IA;

import jakarta.servlet.http.HttpServletRequest;
import net.latinus.sistema.integral.gestion.seguridad.model.both.BodyEncriptado;
import net.latinus.sistema.integral.gestion.seguridad.model.both.ia.ActaExternamientoDTO;
import net.latinus.sistema.integral.gestion.seguridad.model.response.PaginacionResponse;
import net.latinus.sistema.integral.gestion.seguridad.model.response.RespuestaPorDefectoAuditoria;
import org.springframework.web.multipart.MultipartFile;

public interface ActaExternamientoService {

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data ActaExternamientoDTO
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado request peticion.
     * @return RespuestaPorDefectoAuditoria<ActaExternamientoDTO>
     */
    RespuestaPorDefectoAuditoria<PaginacionResponse<ActaExternamientoDTO>> obtenerActasExternamiento(HttpServletRequest httpServletRequest,
                                                                                                           BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean
     * encuesta se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto ActaExternamientoDTO.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> crearActaExternamiento(HttpServletRequest httpServletRequest,
                                                                      BodyEncriptado bodyEncriptado);


    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean
     * empresa se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto SeguimientoPsicologicoDTO.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> actualizarActaExternamiento(HttpServletRequest httpServletRequest,
                                                                           BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean si el documento se sube con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado         objeto BodyEncriptado.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> subirActaFirmada(HttpServletRequest httpServletRequest,
                                                              MultipartFile multipartFile,
                                                              BodyEncriptado bodyEncriptado);

    /**
     * Devuelve un objeto RespuestaPorDefectoAuditoria con data Boolean
     * empresa se creo con exito
     *
     * @param httpServletRequest request peticion.
     * @param bodyEncriptado        objeto SeguimientoPsicologicoDTO.
     * @return RespuestaPorDefectoAuditoria<Boolean>
     */
    RespuestaPorDefectoAuditoria<Boolean> eliminarActaExternamiento(HttpServletRequest httpServletRequest,
                                                                         BodyEncriptado bodyEncriptado);
}

File: pronacej-front/src/app/modules/administracion/services/actaExternamiento.service.ts
import { Injectable } from '@angular/core';
import { BodyEncriptado } from 'app/core/model/both/bodyEncriptado.model';
import { ActaExternamientoDTO } from 'app/core/model/both/ia/actaExternamientoDTO.model';
import { PaginacionRequest } from 'app/core/model/request/PaginacionRequest.model';
import { PaginacionResponse } from 'app/core/model/response/PaginacionResponse.model';
import { RespuestaPorDefecto } from 'app/core/model/response/RespuestaPorDefecto.model';
import { BackendService } from 'app/core/services/backend.service';
import { Observable, Subscriber } from 'rxjs';

@Injectable({
    providedIn: 'root',
})
export class ActaExternamientoService {
    private path = '/actaExternamiento';

    constructor(private backendService: BackendService) {}

    obtenerActasExternamiento(
        paginacionRequest: PaginacionRequest,
        nemonicoMenu: string
    ): Observable<
        RespuestaPorDefecto<PaginacionResponse<ActaExternamientoDTO>>
    > {
        let endPoint = this.path + '/obtenerActasExternamiento';
        return this.backendService.postFinal(
            endPoint,
            paginacionRequest,
            nemonicoMenu
        );
    }

    crearActaExternamiento(
        psicologicoDTO: ActaExternamientoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/crearActaExternamiento';
        return this.backendService.postFinal(
            endPoint,
            psicologicoDTO,
            nemonicoMenu
        );
    }

    actualizarActaExternamiento(
        psicologicoDTO: ActaExternamientoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/actualizarActaExternamiento';
        return this.backendService.postFinal(
            endPoint,
            psicologicoDTO,
            nemonicoMenu
        );
    }

    /**
     * Sube un acta firmada
     *
     * @param actaExternamientoDTO ActaExternamientoDTO
     * @param archivo File
     * @param nemonicoMenu string nemonico menu
     *
     * @return Observable<RespuestaPorDefecto<Boolean>>
     */
    subirActaFirmada(
        actaExternamientoDTO: ActaExternamientoDTO,
        archivo: File,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/subirActaFirmada';
        return new Observable(
            (subs: Subscriber<RespuestaPorDefecto<Boolean>>) => {
                let formData = new FormData();
                formData.append('documento', archivo);
                this.backendService
                    .crearBodyEncriptado(actaExternamientoDTO)
                    .then((bodyEncriptado) => {
                        formData.append('body', JSON.stringify(bodyEncriptado));

                        this.backendService
                            .postFormDataBodyEncriptado2(
                                endPoint,
                                formData,
                                nemonicoMenu
                            )
                            .subscribe({
                                next: async (body: BodyEncriptado) => {
                                    let resp =
                                        await this.backendService.desencriptarBdyEncriptado<
                                            RespuestaPorDefecto<Boolean>
                                        >(body);
                                    subs.next(resp);
                                    subs.complete();
                                },
                                error: (error: any) => {
                                    subs.error(error);
                                    subs.complete();
                                },
                            });
                    })
                    .catch((error: any) => {
                        subs.error(error);
                        subs.complete();
                    });
            }
        );
    }

    eliminarActaExternamiento(
        psicologicoDTO: ActaExternamientoDTO,
        nemonicoMenu: string
    ): Observable<RespuestaPorDefecto<Boolean>> {
        let endPoint = this.path + '/eliminarActaExternamiento';

        return this.backendService.postFinal(
            endPoint,
            psicologicoDTO,
            nemonicoMenu
        );
    }
}
